### PR TITLE
[Profiler] Improve CodeHotspots

### PR DESCRIPTION
**Pull request by gleocadie**
## Summary of changes

Improve Code Hotspots feature.

## Reason for change

We got feedback from customer where spans lasting for 100ms did not have associated callstacks.

## Implementation details

There are 2 main changes:
- Algorithm: we maintain a separate list of threads that has/had a trace context. So each iteration we start collecting threads from this list first (checking that they currently have trace context). By default we will collect 20 threads, this will incur a 4ms-overhead but allows us to reduce the chance to lose 100ms requests.
Then we collect 5 threads (like before) that do not currently have a trace context attached.

- Technical: Before we were using `std::this_thread::sleep_for` but this sleep api does not use a high resolution clock, so we ended up sleeping for longer than expected: For example, we asked to sleep for 9ms, it turned out that we were sleeping for 115ms. Instead we will use `::Sleep` on windows and `usleep` on Linux

## Test coverage

Add a new test where we have only one thread that handles request and 500 idle threads.

### CODE REVIEW

**Review by chrisnas — APPROVED** [review]
See comments

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSamplerLoop.cpp [hunk 1]
```diff
@@ -240,12 +270,58 @@ void StackSamplerLoop::CpuProfilingIteration(void)
...
+
+        if (firstThread == nullptr)
+        {
+            firstThread = _targetThread.get();
+        }
+
+        // skip if it has no trace context
+        if (!_targetThread->HasTraceContext())
+        {
+            continue;
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSamplerLoop.cpp:311]: ✔️

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSamplerLoop.cpp [hunk 2]
```diff
@@ -240,12 +270,58 @@ void StackSamplerLoop::CpuProfilingIteration(void)
...
+        // so prefer bailing out
+        if (_targetThread == nullptr)
+        {
+            break;
+        }
+
+        // keep track of the first seen thread, to avoid infinite loop
+        if (firstThread == _targetThread.get())
+        {
+            break;
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSamplerLoop.cpp:299]: ✔️

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSamplerLoop.cpp [hunk 3]
```diff
@@ -176,30 +179,57 @@ void StackSamplerLoop::MainLoopIteration(void)
...
+        }
 
-            // @ToDo: Investigate whether the OpSysTools::StartPreciseTimerServices(..) invocation made by
-            // the StackSamplerLoopManager ctor really ensures that this yield for 1ms or less.
-            // If not, we should not be yielding here.
-            std::this_thread::yield();
+        // skip thread if it has a trace context
+        if (_targetThread->HasTraceContext())
+        {
+            continue;
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSamplerLoop.cpp:217]: ✔️

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSamplerLoop.cpp [hunk 4]
```diff
@@ -176,30 +179,57 @@ void StackSamplerLoop::MainLoopIteration(void)
...
-            int64_t thisSampleTimestampNanosecs = OpSysTools::GetHighPrecisionTimestamp();
-            int64_t prevSampleTimestampNanosecs = _targetThread->SetLastSampleHighPrecisionTimestampNanoseconds(thisSampleTimestampNanosecs);
-            int64_t duration = ComputeWallTime(thisSampleTimestampNanosecs, prevSampleTimestampNanosecs);
+            break;
+        }
 
-            CollectOneThreadStackSample(_targetThread, thisSampleTimestampNanosecs, duration, PROFILING_TYPE::WallTime);
+        if (firstThread == _targetThread.get())
+        {
+            break;
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSamplerLoop.cpp:205]: ✔️

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp [hunk 5]
```diff
@@ -366,6 +372,13 @@ int32_t Configuration::ExtractWallTimeThreadsThreshold()
     return threshold;
 }
 
+int32_t Configuration::ExtractCodeHotspotsThreadsThreshold()
+{
+    // default threads to sample for codehotspots is 20; could be changed via env vars from 5 to 64
+    int32_t threshold = std::max(GetEnvironmentValue(EnvironmentVariables::CodeHotspotsThreadsThreshold, 20), 1);
```
[the remark this answers is not in the document]
Comment (gleocadie) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/Configuration.cpp]: ✔️

**Pull request merged**